### ISSUE TITLE
Feature request: Parser, SNS envelope: Add messageId to parsed event

### ISSUE THREAD

**Issue by Twonky-Chris** [aws-powertools/powertools-lambda-typescript#3323]
### Use case

The Parser returns only the parsed Message. It would be very helpful to also have the messageId in hand, for example, for error handling, especially as item identifier for batchItemFailures.

### Solution/User Experience

maybe like this:

.use(
            parser({ schema: validator, envelope: SnsSqsEnvelope, options: {returnMessageId: true })
        )

### Alternative solutions

```Markdown

```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [x] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

**Comment by dreamorosi** [comment]
Hi @Twonky-Chris, thank you for opening this issue.

Just to clarify your use case, in your code snippet `validator` is a custom Zod schema that you're using to specify how we should parse the `message` field in each SNS message, is that correct?

**Comment by Twonky-Chris** [comment]
> Hi [@Twonky-Chris](https://github.com/Twonky-Chris), thank you for opening this issue.
> 
> Just to clarify your use case, in your code snippet `validator` is a custom Zod schema that you're using to specify how we should parse the `message` field in each SNS message, is that correct?

Yes, correct. My suggestion was just about the "options: {returnMessageId: true }".

**Comment by everett1992** [comment]
I think there's two usecases to consider and we could handle them with different methods. The first is that you just want to access the envelop's content, and  the second is that you want the evelope and content.

```
const event = { Records: [ { messageId: "...", body: "{\"foo\": 1}" } ] }
const schema = z.object({foo: z.string()})
type Schema = z.infer<typeof schema>

// Just get the array of parsed message bodies
SqsSnsEnvelope.parseInner(event, schema) satisfies Schema[]
 => [ { foo: 1 } ]

// get the array of parsed message bodies and keep the envelope.
// Not sure what the type should be 
// - should we keep the Records wrapper, or is there no value there?
// - should we keep the unparsed body so we still have a SQSRecord, or replace it?
SqsSnsEnvelope.parseAll(event, schema) satisfies { Records: (SqsSnsMetadata & Schema)[] }
 => { Records: [ { messageid: "...", body: { foo: 1 } }] }
```

**Comment by dreamorosi** [comment]
Thank you both for the context.

I agree that right now, for certain/most of the built-in schemas and envelopes there's this grey area between just using the built-in schema and using the corresponding built-in envelope.

Right now you can either use the schema, which gives you access to the entire parsed event but without transforming the payload, or you use the envelope which does transform the payload into something usable but locks you out of all other fields.

Our initial assumption was that most customers would only be interested in the payload (i.e. `message`, `body`, etc.) so envelopes fit the requirements, and for all other customers who needed more fields, schemas were available. This is probably not 100% accurate.

We have already started discussing how to solve this, and generally speaking we are not likely going to add extra options like `options: {returnMessageId: true }`. If we did that, next week or next month another customer could come and ask us to add a new option for another field, and eventually we could end up redefining the entire schema in the options - which is not where we want to go.

Adding new custom methods on the envelope like `parseInner()` and `parseAll()` is an option. In the design for envelopes we tried to stay as close as possible to the Zod API instead of adding new methods. This helps with familiarity but also keeps the door open to add support for other current or future Zod features like `parseAsync()`.

We have some ideas and @am29d is working on this - so we expect to be able to share more in the coming weeks.

**Comment by Twonky-Chris** [comment]
Thank you for the fast feedback. Now that I am thinking about it, ZOD could also be used to define the entire object, including attributes from the envelope, that should be returned. I am looking forward to your ideas.

**Comment by dreamorosi** [comment]
Hi @Twonky-Chris, thank you for your patience here.

We have been discussing changes to Parser internally for a while and eventually we settled on two points: 1/ built-in schemas & envelopes should have defined boundaries, 2/ we won't implement custom methods on envelopes at this time.

The first point boils down to the fact that we offer both built-in schemas (i.e. `SqsSchema`) and envelopes (i.e. `SqsEnvelope`). Envelopes build on top of schemas and their single purpose is to provide syntactic sugar for extracting the payload field from a schema while parsing it against a customer-provided schema.

As a rule of thumb, if you need only the inner payload of an event, for example the `body` in a SQS message, you should use an envelope, otherwise you should use a schema.

Schemas provide greater flexibility and besides giving you access to the entire parsed event, also allow you to extend them with your own schemas, essentially achieving the same goal.

In your case, since you want to have both the SNS Notification within the body and the `messageId`, you should probably reach for a schema rather than an envelope.

Over the last couple months we have also added Zod helpers like `JSONStringified` that apply similar transforms as the ones used in the envelopes, and that we think should make these use cases less verbose.

Below is an example of your use case with both Parser and Batch Processing, with the record handler function having access to the entire SQS record message already parsed, as well as the inner `body` (aka the SNS Notification) transformed and parsed using a custom schema:

```ts
import middy from '@middy/core';
import { parser } from '@aws-lambda-powertools/parser/middleware';
import { z } from 'zod';
import { SnsSqsNotificationSchema } from '@aws-lambda-powertools/parser/schemas/sns';
import {
  SqsSchema,
  SqsRecordSchema,
} from '@aws-lambda-powertools/parser/schemas/sqs';
import { JSONStringified } from '@aws-lambda-powertools/parser/helpers';
import {
  BatchProcessor,
  EventType,
  processPartialResponse,
} from '@aws-lambda-powertools/batch';
import { Logger } from '@aws-lambda-powertools/logger';

const logger = new Logger({ logLevel: 'DEBUG' });
const batchProcessor = new BatchProcessor(EventType.SQS);

// Schema of the Message field in your SNS notification
const messageSchema = z
  .object({
    name: z.string(),
    age: z.number(),
  })
  .strict();

// Schema of the SQS record
const sqsRecordSchema = SqsRecordSchema.extend({
  body: JSONStringified(
    SnsSqsNotificationSchema.extend({
      Message: JSONStringified(messageSchema),
    })
  ),
});

// Schema of the event
const eventSchema = SqsSchema.extend({
  Records: z.array(sqsRecordSchema).min(1),
});

// The record handler has access to the parsed & transformed records
const recordHandler = async (record: z.infer<typeof sqsRecordSchema>) => {
  const {
    body: {
      Message: { name, age }, // this is already a parsed/transformed object
    },
    messageId,
  } = record;

  logger.debug('processing message', {
    messageId,
    user: {
      name,
      age,
    },
  });
};

export const handler = middy()
  .use(
    parser({
      schema: eventSchema,
    })
  )
  .handler(async (event, context) =>
    processPartialResponse(event, recordHandler, batchProcessor, {
      context,
    })
  );
```

While I acknowledge that the experience is not as terse as the one using envelopes, we have decided to go this way because we want to avoid creating a wide-surface API that allows you to use envelopes but selectively getting other fields. We think this would be confusing and would eventually result in a sprawl since we'd have to account for dozens of edge cases for each event source.

Ultimately if there's a need for greater flexibility, it's not as farfetched to have a little more complexity. 

---

With this in mind, I'll be closing the issue as not planned.
